### PR TITLE
feat: initial testing of lockfile and coroutine onexit implementation

### PR DESCRIPTION
**Pull request by khris190**
so this is initial version of lockfile on_exit implementation, it seems to work i just want to share the progress
TODO:

- i'll move lock_file functions to utils and parametrize them
- remove dump from 3rd line
- remove temporary sleep 5 command
- remove special treatment of that full_cmd and implement on_exit properly
- tests

if i'm missing something then please tell me

### CODE REVIEW

**Comment by Civitasv** [comment]
Seems right to me.

**Comment by khris190** [comment]
i added hardcoded 0 to on_exit() in terminal for now, but i'm thinking that we could store command exit code in another file so we could get it after

**Comment by khris190** [comment]
also why don't runners have any callbacks?

**Comment by Civitasv** [comment]
Executors are used for cmake-rated commands, including generate, build, install, clean, etc. Runners are used for running executables. Now the end of this plugin is Runner.

**Comment by khris190** [comment]
i think it might be complete. i'll use it a bit so i'm sure everything works as intended

**Comment by Civitasv** [comment]
> i'm also wondering if calling those files "tmp" is okay.

I think we should call it `cmake_tools_lockfile`

File: lua/cmake-tools/utils.lua [hunk 1]
```diff
@@ -61,6 +61,22 @@ function utils.file_exists(path)
...
+---@return string
+function utils.get_tmp_file_path(file_name)
+  return utils.get_tmp_dir() .. "/" .. file_name
+end
+
+function utils.create_tmp_file(file_name)
+  utils.mkdir(utils.get_tmp_dir())
+  local tmp_file = io.open(utils.get_tmp_file_path(file_name), "w")
+  tmp_file:close()
+end
```
Comment (Civitasv) [inline lua/cmake-tools/utils.lua:79]: These functions are specific to `terminal`, should be moved to there.

File: lua/cmake-tools/terminal.lua [hunk 2]
```diff
@@ -611,5 +668,16 @@ end
 function _terminal.is_installed()
   return true
 end
-
+--- Handle when a terminal process exists
+---@param opts any
+---@param on_exit function|nil function to be executed on exit
+---@param close_on_exit boolean close the terminal on exit
+function _terminal.__handle_exit(opts, on_exit, close_on_exit)
```
Comment (Civitasv) [inline lua/cmake-tools/terminal.lua:675]: why not just name it to `handle_exit`?

File: lua/cmake-tools/terminal.lua [hunk 3]
```diff
@@ -527,9 +530,52 @@ function _terminal.prepare_cmd_for_execute(cmd, env, args)
...
-  local prefix = opts.prefix_name -- [CMakeTools]
+local get_lock_file_path = function()
+  return utils.get_tmp_file_path("commandRunning.lock")
+end
+
+local get_last_exit_code_file_path = function()
+  return utils.get_tmp_file_path("lastCmdExitCode.txt")
+end
+
+local create_tmp_lock_file = function()
```
Comment (Civitasv) [inline lua/cmake-tools/terminal.lua:541]: create_lock_file is okay

File: lua/cmake-tools/terminal.lua (hunk 2 above)
Comment (khris190) [reply lua/cmake-tools/terminal.lua:675]: i copied it from toggleterm without thinking, i now see that it doesn't work with our naming scheme

File: lua/cmake-tools/utils.lua (hunk 1 above)
Comment (khris190) [reply lua/cmake-tools/utils.lua:79]: my idea was that the tmp directory handling is not specyfic to terminal, it is just that only terminal uses it for now, but if we ever need any other tmp file stored we could use the same functions

Comment (Civitasv) [reply lua/cmake-tools/utils.lua:79]: I still think they are too specific to be in utils.

**Comment by Civitasv** [comment]
LGTM. Thanks for your fantastic work!

**Pull request merged**